src/connector/src/source/external.rs

```diff
@@ -0,0 +1,622 @@
+    }
+
+    fn current_cdc_offset(&self) -> Self::CdcOffsetFuture<'_> {
+        async move {
+            let mut conn = self
+                .pool
+                .get_conn()
+                .await
+                .map_err(|e| ConnectorError::Connection(anyhow!(e)))?;
+
+            let sql = "SHOW MASTER STATUS".to_string();
+            let mut rs = conn.query::<mysql_async::Row, _>(sql).await?;
+            let row = rs
+                .iter_mut()
+                .exactly_one()
+                .map_err(|e| ConnectorError::Internal(anyhow!("read binlog error: {}", e)))?;
+
+            Ok(CdcOffset::MySql(MySqlOffset {
+                filename: row.take("File").unwrap(),
+                position: row.take("Position").unwrap(),
+            }))
+        }
+    }
+
+    fn parse_binlog_offset(&self, offset: &str) -> ConnectorResult<CdcOffset> {
+        Ok(CdcOffset::MySql(MySqlOffset::parse_str(offset)?))
+    }
+
+    fn snapshot_read(
+        &self,
+        table_name: SchemaTableName,
+        start_pk: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) -> BoxStream<'_, ConnectorResult<OwnedRow>> {
+        self.snapshot_read_inner(table_name, start_pk, primary_keys)
+    }
+}
+
+impl MySqlExternalTableReader {
+    pub fn new(properties: HashMap<String, String>, rw_schema: Schema) -> ConnectorResult<Self> {
+        if let Some(field) = rw_schema.fields.last() && field.name.as_str() != OFFSET_COLUMN_NAME {
+            return Err(ConnectorError::Config(anyhow!(
+                "last column of schema must be `_rw_offset`"
+            )));
+        }
+
+        let config = serde_json::from_value::<ExternalTableConfig>(
+            serde_json::to_value(properties).unwrap(),
+        )
+        .map_err(|e| {
+            ConnectorError::Config(anyhow!("fail to extract mysql connector properties: {}", e))
+        })?;
+
+        let database_url = format!(
+            "mysql://{}:{}@{}:{}/{}",
+            config.username, config.password, config.host, config.port, config.database
+        );
+        let pool = mysql_async::Pool::from_url(database_url)?;
+
+        let field_names = rw_schema
+            .fields
+            .iter()
+            .filter(|f| f.name != OFFSET_COLUMN_NAME)
+            .map(|f| f.name.as_str())
+            .join(",");
+
+        Ok(Self {
+            pool,
+            config,
+            rw_schema,
+            field_names,
+        })
+    }
+
+    pub async fn disconnect(&self) -> ConnectorResult<()> {
+        self.pool.clone().disconnect().await?;
+        Ok(())
+    }
+
+    #[try_stream(boxed, ok = OwnedRow, error = ConnectorError)]
+    async fn snapshot_read_inner(
+        &self,
+        table_name: SchemaTableName,
+        start_pk: Option<OwnedRow>,
+        primary_keys: Vec<String>,
+    ) {
+        let order_key = primary_keys.iter().join(",");
+        let sql = if start_pk.is_none() {
+            format!(
+                "SELECT {} FROM {} ORDER BY {}",
```

Comment:
Could there be potential injection attacks here?

Reply:
> Could there be potential injection attacks here?

Are spaces allowed in the field name?

Reply:
It seems possible.

```
dev=> create table t ("1; DROP DATABASE foo; SELECT 1" int);
CREATE_TABLE
dev=> select * from t;
 1; DROP DATABASE foo; SELECT 1 
--------------------------------
(0 rows)
```

Similar issues: https://github.com/risingwavelabs/risingwave/issues/4217

Reply:
Thanks for the reminder🥵

Reply:
I will unify to binary protocol in a separation PR.